[Ma Yuchen] iP

dawg420 opened the pull request:
# Duke XD 🤣

Duke XD is a cutting-edge to-do list app that makes life much easier:
- Keep track of deadlines
- Mark tasks as done/undone
- ~~???~~

All you need to do is:
1. Download it from [here](https://github.com/dawg420/ip).
2. Enter 'java Main.java' in terminal
3. ???
4. Profit!

Features:
- [x] Date time functionality
- [x] Search functionality for tasks
- [ ] Minecraft minigame

Some sample commands:
```
todo buy book
deadline return book /by 2023-03-15 00:00
list
```

> Sample text

Commit:
Add Gradle support

dawg420 committed:
Level 1: Added greet, echo, exit

dawg420 committed:
Level 2: Added add, list

dawg420 committed:
Level 3: Added mark, unmark functions

dawg420 committed:
Level 4: Added Todos, Events, Deadlines

dawg420 committed:
Used i/o redirection technique to semi-automate the testing of Duke

dawg420 committed:
Handles errors when the task to mark is invalid or outside of the list; when the task is not preceded by the type of the task; and when the todo field is empty

dawg420 committed:
Added functionality to delete tasks

dawg420 committed:
A-Enums

Added A-Enums for command type

dawg420 committed:
Level 7. Save

Added Save functionality

dawg420 committed:
Level-7: Added save functionality

dawg420 committed:
Level 8.

Added Date Time functionality for deadlines

dawg420 committed:
Merge branch 'branch-Level-7'

dawg420 committed:
A-MoreOOP

Refactored the code to extract out closely related code as classes.

dawg420 committed:
A-Packages

Put all classes into a duke package

dawg420 committed:
Updated with Intellij

Switched to Intellij instead of VSCode

dawg420 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

dawg420 committed:
A-Gradle

Set up gradle

dawg420 committed:
A-JUnit

Added 2 J-unit tests

dawg420 committed:
A-JavaDoc

Add JavaDoc comments to the code

dawg420 committed:
A-CodingStandard

Tweaked the code to comply with given coding standard.

dawg420 committed:
Level-9

Added find functionality

dawg420 committed:
Merge branch 'branch-A-CodingStandard'

dawg420 committed:
Merge branch 'branch-Level-9'

dawg420 committed:
Bug fixes

lyndonlim27 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+		String toWhen = "";
+
+		switch (cmdType) {
+		case MARK:
+			try {
+				firstWord = msg.substring(0, msg.indexOf(" "));
+				secondInt = Integer
+						.parseInt(msg.substring(msg.indexOf(" ") + 1, msg.length()));
+				tasklist.mark(secondInt);
+				try {
+					Storage.save(tasklist);
+				} catch (Exception ex) {
+					System.err.println("Error! There is no save file!");
+				}
+			} catch (Exception ex) {
+				System.err.println("Please indicate a valid task!");
+			}
+			break;
```
Avoid Deep Nesting

lyndonlim27 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+					Storage.save(tasklist);
+				} catch (Exception ex) {
+					System.err.println("Error! There is no save file!");
+				}
+			} catch (Exception ex) {
+				System.err.println("Whoops! Please enter the deadline followed by its due date preceeded by a '/by'." +
+						"The date time format should be yyyy-mm-dd hh:mm");
+			}
+			break;
+		case EVENT:
+			try {
+				firstWord = msg.substring(msg.indexOf(" ") + 1, msg.indexOf("/from") - 1);
+				fromWhen = msg.substring(msg.indexOf("/from") + 6, msg.indexOf("/to") - 1);
+				toWhen = msg.substring(msg.indexOf("/to") + 4, msg.length());
+				Task.Event newEvent = new Task.Event(firstWord, fromWhen, toWhen);
+				tasklist.add(newEvent);
+				System.out.println("Got it. I've added this task:");
+				System.out.println(newEvent);
+				System.out.println("Now you have " + tasklist.size() + " tasks in the list.");
+				try {
+					Storage.save(tasklist);
+				} catch (Exception ex) {
+					System.err.println("Error! There is no save file!");
+				}
+			} catch (Exception ex) {
+				System.err.println("Whoops! Please enter the event followed by its /from and /to timings.");
+			}
+			break;
+		case BYE:
+			Ui.showGoodbyeMessage();
+			echoScanner.close();
+			loop = false;
+			break;
+		case LIST:
+			tasklist.printList();
+			break;
+		}
+		return loop;
+	}
+}
```
Avoid Long Methods

> Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle. [(Source: Avoid long methods)](https://nus-cs2103-ay2223s2.github.io/website/se-book-adapted/chapters/codeQuality.html)

lyndonlim27 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+
+		@Override
+		public String toString() {
+			return "[T]" + super.toString();
+		}
+	}
+
+	public static class Deadline extends Task {
+		protected LocalDate by;
+		protected LocalTime when;
+		DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd, yyyy");
+
+		public Deadline(String description, LocalDate by, LocalTime when) {
+			super(description);
+			this.by = by;
+			this.when = when;
+		}
+
+		@Override
+		public String toString() {
+			return "[D]" + super.toString() + " (by: " + formatter.format(by) + " " + when + ")";
+		}
+	}
+
+	public static class Event extends Task {
+		protected String from;
+		protected String to;
+
+		public Event(String description, String from, String to) {
+			super(description);
+			this.from = from;
+			this.to = to;
+		}
+
+		@Override
+		public String toString() {
+			return "[E]" + super.toString() + " (from: " + from + " to: " + to + ")";
+		}
+	}
+}
```
Missing javadocs

lyndonlim27 commented on src/main/java/duke/Tasklist.java:
```diff
@@ -0,0 +1,77 @@
+ */
+
+public class Tasklist {
+	ArrayList<Task> listOfThings = new ArrayList<Task>();
```
Great! You distinguish clearly between single-valued and multi-valued variables.

daytona65 replied:
Great job on variable naming between single valued and multi valued variables

lyndonlim27 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,41 @@
+	 * Prints the welcome message whenever
+	 * the program is started
+	 */
+	public static void showWelcomeMessage() {
+		String welcomeMsg = "Hello from\n" + logo +
+				"\n" + "Hello! I'm Duke\nWhat can I do for you?";
+		System.out.println(welcomeMsg);
+	}
+
+	/**
+	 * Prints the goodbye message whenever
+	 * the program is terminated.
+	 */
+	public static void showGoodbyeMessage() {
+		System.out.println("Bye. Hope to see you again soon!");
+	}
```
Great function names! You explain the named entity to readers accurately and at a sufficient level of detail.

lyndonlim27 reviewed (commented):
Well done! You followed most of the code quality guidelines except for a few I pointed out! Your javadocs were done well too!

daytona65 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,90 @@
+			String taskName = "";
+
+
+			switch (taskType) {
```
I like the use of switches here

daytona65 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+
+public class Parser {
+
+	enum Type {
```
I like the use of enum here

daytona65 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,76 @@
+	protected String description;
+	protected boolean isDone;
+
+	public Task(String description) {
+		this.description = description;
+		this.isDone = false;
+	}
+
+	public String getStatusIcon() {
+		return (isDone ? "X" : " ");
+	}
+
+	public void setDone() {
+		this.isDone = true;
+	}
+
+	public void setUndone() {
+		this.isDone = false;
```
Might want to add javadocs

daytona65 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,41 @@
+	 * Prints the welcome message whenever
+	 * the program is started
+	 */
+	public static void showWelcomeMessage() {
```
Method naming is correct

daytona65 reviewed (commented):
Overall great, clean code that largely follows coding standards aside from some missing Javadocs. Great work!

dawg420 committed:
Level-10

Implemented a GUI (Most basic)

dawg420 committed:
OOP fixes

Extracted out commands and implemented more SLAP.

dawg420 committed:
CodeQuality

Storage class now has private functions loadTodo, loadDeadline and loadEvent

Extracting these functions allows the code to follow the SLAP principle more closely.

This is done to ensure that different levels of abstraction are contained in different blocks of code, keeping the code clean.

dawg420 committed:
Merge pull request #2 from dawg420/branch-Code-Quality

CodeQuality

dawg420 committed:
Implemented Assertions

Incorporated assertions to the storage class

Manipulation of the save file during the running of the application could lead to unforeseen errors.

Used assertions to make sure that the save file exists at the specified location all along the way

This makes the code less error prone.

dawg420 committed:
Streams implementation

Used streams to implement the search function of tasks in the tasklist

Eliminates the use of for-loops, increases efficiency

dawg420 committed:
Merge pull request #3 from dawg420/branch-A-Assertions

Implemented Assertions

dawg420 committed:
Merge branch 'master' into A-Streams

dawg420 committed:
Merge pull request #4 from dawg420/A-Streams

Streams implementation

dawg420 committed:
Implemented multiple saves

yes

dawg420 committed:
Improved Code Quality

Extracted out some functions to make the function body shorter. Doing so makes the code more readable.

Rewrote Javadocs for all the classes and methods to follow the coding convention.

dawg420 committed:
Bug fixes

Fixed an annoying bug.

dawg420 committed:
Merge branch 'master' into branch-CodeQualityFixes

dawg420 committed:
Update README.md

Updated the user guide

dawg420 committed:
Update README.md

dawg420 committed:
Updated readme

dawg420 committed:
Update README.md

dawg420 committed:
Update README.md

dawg420 committed:
Uploaded Ui.jpg

dawg420 committed:
Update README.md

dawg420 committed:
Update README.md

dawg420 committed:
Update README.md

dawg420 committed:
Update README.md

dawg420 committed:
Bug fixes

dawg420 committed:
Update README.md

dawg420 committed:
Updated readme

dawg420 committed:
Update README.md

dawg420 committed:
Update README.md

dawg420 committed:
Add files via upload

dawg420 committed:
Delete Ui.jpg

dawg420 committed:
Update README.md
